server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -1842,6 +1842,164 @@ public void testDotsOnlyFieldNames() throws Exception {
         assertThat(err.getCause().getMessage(), containsString("field name cannot contain only dots"));
     }
 
+    public void testCollapsedObjectWithInnerObject() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(
+            mapping(b -> { b.startObject("metrics.service").field("type", "object").field("collapsed", true).endObject(); })
+        );
+        IllegalArgumentException err = expectThrows(IllegalArgumentException.class, () -> mapper.parse(source("""
+            {
+              "metrics": {
+                "service": {
+                  "time" : {
+                    "max" : 10
+                  }
+                }
+              }
+            }
+            """)));
+        assertEquals("Object [metrics.service] is collapsed and does not support inner object [time]", err.getMessage());
+    }
+
+    public void testCollapsedObjectWithInnerDottedObject() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(
+            mapping(b -> { b.startObject("metrics.service").field("type", "object").field("collapsed", true).endObject(); })
+        );
+        IllegalArgumentException err = expectThrows(IllegalArgumentException.class, () -> mapper.parse(source("""
+            {
+              "metrics": {
+                "service": {
+                  "test.with.dots" : {
+                    "max" : 10
+                  }
+                }
+              }
+            }
+            """)));
+        assertEquals("Object [metrics.service] is collapsed and does not support inner object [test.with.dots]", err.getMessage());
+    }
+
+    public void testCollapsedRootWithInnerObject() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> b.field("collapsed", true)));
+        IllegalArgumentException err = expectThrows(IllegalArgumentException.class, () -> mapper.parse(source("""
+            {
+              "metrics": {
+                "service": {
+                  "time.max" : 10
+                }
+              }
+            }
+            """)));
+        assertEquals("Object [_doc] is collapsed and does not support inner object [metrics]", err.getMessage());
```

Comment:
I wonder if we can detect this case and replace `Object [_doc]` with `root`, or something else?  It's a bit confusing otherwise, as there isn't a `_doc` anywhere in the mappings that the user sees.

Reply:
It would be nice to get the full object path here as well, but from a quick look at the parsing code it doesn't look like it will be simple to do, so happy to leave both of these as a TODO.

Reply:
yea I agree about the error message, I believe that's what we already do for enabled though. I would look at changing this as a follow-up.

Which full object path do you mean?